src/qt/editor/undomanager.cpp

```diff
@@ -43,23 +43,99 @@
 #include <QGuiApplication>
 #include <warn/pop>
 
+#include <thread>
+#include <mutex>
+#include <condition_variable>
+#include <atomic>
+#include <vector>
+#include <string>
+
 namespace inviwo {
 
+class AutoSaver {
+public:
+    AutoSaver()
+        : path_{filesystem::getPath(PathType::Settings)}
+        , restored_{[this]() -> std::optional<std::string> {
+            if (filesystem::fileExists(path_ + "/autosave.inv")) {
+                auto ifstream = filesystem::ifstream(path_ + "/autosave.inv");
+                std::stringstream buffer;
+                buffer << ifstream.rdbuf();
+                return std::move(buffer).str();
+            }
+
+            return std::nullopt;
+        }()}
+        , quit_{false}
+        , saver_{[this]() {
+            for (;;) {
+
+                std::shared_ptr<const std::string> str;
+                {
+                    std::unique_lock<std::mutex> lock(mutex_);
+                    while (!quit_ && toSave_.empty()) {
+                        condition_.wait(lock);
+                    }
+
+                    if (quit_) return;
+
+                    if (!toSave_.empty()) {
+                        str = std::move(toSave_.back());
+                        toSave_.clear();
+                    }
+                }
+
+                if (str) {
+                    auto ofstream = filesystem::ofstream(path_ + "/autosave.inv.tmp");
+                    ofstream << *str;
+                    filesystem::copyFile(path_ + "/autosave.inv.tmp", path_ + "/autosave.inv");
+                }
+            }
+        }} {}
+
+    ~AutoSaver() {
+        quit_ = true;
+        condition_.notify_one();
+        saver_.join();
+    }
+
+    const std::optional<std::string> &getRestored() const { return restored_; }
+
+    void save(std::shared_ptr<const std::string> str) {
+        {
+            std::unique_lock<std::mutex> lock(mutex_);
+            toSave_.push_back(str);
+        }
+        condition_.notify_one();
+    }
+
+private:
+    std::string path_;
+    std::optional<std::string> restored_;
+    std::atomic<bool> quit_;
+    std::condition_variable condition_;
+    std::mutex mutex_;
+    std::vector<std::shared_ptr<const std::string>> toSave_;
+
+    std::thread saver_;
+};
+
 UndoManager::UndoManager(InviwoMainWindow *mainWindow)
     : mainWindow_(mainWindow)
     , manager_{mainWindow_->getInviwoApplication()->getWorkspaceManager()}
-    , refPath_{filesystem::findBasePath()} {
+    , refPath_{filesystem::findBasePath()}
+    , autoSaver_{std::make_unique<AutoSaver>()} {
 
     mainWindow_->getInviwoApplicationQt()->setUndoTrigger([this]() { pushStateIfDirty(); });
```

Comment:
Would be better if we could use the app directly to avoid dependency on IniwoMainWindow.

Reply:
The window is used in a few more places.... guess I can disconnect them, but until I have a concrete use case, that would be premature generalization

Reply:
I now have an app that could use it :) 
but sure, it is another PR